src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2585 @@
+    //   VEC_SIZE one at a 32       to check for early mismatches. Only do this
+    //   if its guaranteed the 33       work is not wasted. 34    5. If size is
+    //   8 * VEC_SIZE or less, unroll the loop. 35    6. Compare 4 * VEC_SIZE at
+    //   a time with the aligned first memory 36       area. 37    7. Use 2
+    //   vector compares when size is 2 * VEC_SIZE or less. 38    8. Use 4
+    //   vector compares when size is 4 * VEC_SIZE or less. 39    9. Use 8
+    //   vector compares when size is 8 * VEC_SIZE or less.  */
+
+    Label L_less_vec, L_return_vec_0, L_last_1x_vec, L_return_vec_1,
+        L_last_2x_vec;
+    Label L_return_vec_2, L_retun_vec_3, L_more_8x_vec, L_return_vec_0_1_2_3;
+    Label L_return_vzeroupper, L_8x_return_vec_0_1_2_3, L_loop_4x_vec;
+    Label L_return_vec_3, L_8x_last_1x_vec, L_8x_last_2x_vec, L_8x_return_vec_2;
+    Label L_8x_return_vec_3, L_return_vec_1_end, L_return_vec_0_end,
+        L_one_or_less;
+    Label L_page_cross_less_vec, L_between_16_31, L_between_8_15, L_between_2_3,
+        L_zero;
+    Label L_ret_nonzero;
+
+    // 72              .section SECTION(.text),"ax",@progbits
+    // 73      ENTRY (MEMCMP)
+    // 74      # ifdef USE_AS_WMEMCMP
+    // 75              shl     $2, %RDX_LP
+    // 76      # elif defined __ILP32__
+    // 77              /* Clear the upper 32 bits.  */
+    // 78              movl    %edx, %edx
+
+    // 79      # endif
+
+    // 80              cmp     $VEC_SIZE, %RDX_LP
+    // 81              jb      L(less_vec)
+    // 82
+    __ cmpq(rdx, 0x20);
+    __ jb(L_less_vec);
+    __ vmovdqu(xmm1, Address(rsi, 0));
+
+    // 83              /* From VEC to 2 * VEC.  No branch when size == VEC_SIZE.
+    // */
+    __ vpcmpeqb(xmm1, xmm1, Address(rdi, 0), Assembler::AVX_256bit);
+
+    // 84              vmovdqu (%rsi), %ymm1
+    __ vpmovmskb(rax, xmm1, Assembler::AVX_256bit);
+
+    // 85              VPCMPEQ (%rdi), %ymm1, %ymm1
+    // 86              vpmovmskb %ymm1, %eax
+    // 87              /* NB: eax must be destination register if going to
+    // 88                 L(return_vec_[0,2]). For L(return_vec_3 destination
+    // register 89                 must be ecx.  */
+    __ incrementl(rax);
+    __ jne(L_return_vec_0);
+
+    // 90              incl    %eax
+    // 91              jnz     L(return_vec_0)
+    __ cmpq(rdx, 0x40);
+
+    // 92
+    __ jbe(L_last_1x_vec);
+
+    // 93              cmpq    $(VEC_SIZE * 2), %rdx
+    // 94              jbe     L(last_1x_vec)
+    // 95
+    __ vmovdqu(xmm2, Address(rsi, 0x20));
+
+    // 96              /* Check second VEC no matter what.  */
+    __ vpcmpeqb(xmm2, xmm2, Address(rdi, 0x20), Assembler::AVX_256bit);
+
+    // 97              vmovdqu VEC_SIZE(%rsi), %ymm2
+    __ vpmovmskb(rax, xmm2, Assembler::AVX_256bit);
+
+    // 98              VPCMPEQ VEC_SIZE(%rdi), %ymm2, %ymm2
+    // 99              vpmovmskb %ymm2, %eax
```

Comment:
It seems that here the comments and code is strangely interleaved / shifted. What is this all for?

Reply:
All this has been remedied